### ISSUE TITLE
Button styling missing on add to cart buttons when using a products shortcode on the homepage

### ISSUE THREAD

**Issue by mikejolley** [woothemes/woocommerce#52907]
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

When using a `[products]` shortcode on the homepage, adding a product to the cart does not show the "check" icon on the button as it does on other store pages.

![Image](https://github.com/user-attachments/assets/75693231-fa38-4816-860e-cf3706d28193)

The styles for this are present here: https://github.com/woocommerce/woocommerce/blob/6f817942161649daa5a16534ae1246e6d3215299/plugins/woocommerce/client/legacy/css/woocommerce.scss#L1710-L1715

It seems to work if you remove the `:where(body:not(.woocommerce-block-theme-has-button-styles))` selector. This class is not present on the body tag so this selector should not fail. 

The only difference I see is that on the pages it works, the `body` has the `woocommerce` classname. On the homepage that doesn't work, the `woocommerce` classname is added around the shortcode only, not on the body tag.

9051098-zd-a8c

### Expected behavior

There should be a "check" icon after a successful add.

### Actual behavior

No styling.

### Steps to reproduce

1. Edit the page assigned as the homepage
2. Add a shortcode block containing `[products]`
3. Visit the page and add an item to the cart
4. See styling

### WordPress Environment

N/A

### Isolating the problem

- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Comment by EmmaHarrison96** [comment]
Hi Mike,
Well, I haven’t personally encountered this exact problem with the WooCommerce plugin’s product shortcode on the homepage, I wanted to mention an alternative solution that worked flawlessly for me. I’ve been using the Custom Add to Cart Button for WooCommerce plugin by Addify ([check it out here](https://addify.store/product/custom-add-to-cart-button-for-woocommerce)) for a similar setup.

With this plugin, the button styling and functionality—including adding products to the cart—worked perfectly without needing additional adjustments to themes or CSS. It’s extremely user-friendly and integrates seamlessly, even on custom layouts like the homepage.

If troubleshooting the existing setup doesn’t resolve the issue, you might find this alternative useful. Hope this helps, and good luck resolving the bug!

**Comment by gigitux** [comment]
I'm unable to see the icon even on WooCommerce pages like the Product Catalog:

https://github.com/user-attachments/assets/84a64e22-c364-49cf-920c-2540b2637848


While I acknowledge that this is a bug, I would not prioritize this issue given that on block themes, there is the Product Collection block built to show the Product Grid - what do you think, @mikejolley?

**Comment by gigitux** [comment]
Moreover, it looks like that the `[products]` short code suffers other layout issues on block themes.

**Comment by mikejolley** [comment]
@gigitux Wouldn't the issue affect all themes though? The problem is caused by that CSS selector above. Could the exclusion rule for `.woocommerce-block-theme-has-button-styles` just be moved into block styles instead of this core stylesheet?

**Comment by octaedro** [comment]
Thank you, @mikejolley, for reporting this.

Just a small detail: as far as I can see, we don't load the [woocommerce.scss](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/legacy/css/woocommerce.scss) file when using one of the [theme-support](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce/includes/theme-support) classes.

That means that the problem with `woocommerce.scss` will only be present in TT4 and TT5. What caught my attention was @gigitux's [example](https://github.com/woocommerce/woocommerce/issues/52907#issuecomment-2493517695), as it seems like he used TT3.

So, when fixing this, we should not forget that scenario.

**Comment by gigitux** [comment]
Sorry folks, can you share how I can get the expected behavior? I'm trying with WooCommerce 9.3 on a Single Product Page and I'm unable to see the "check" icon:

https://github.com/user-attachments/assets/b9c0cfe7-346e-4b1b-8ea0-20edd812ed8f

I'm not able to see it on a classic theme too:

https://github.com/user-attachments/assets/bc288613-cc12-4b2c-b7c9-129d4e6f69a5

**Comment by danielwrobert** [comment]
I'm in the same boat as @gigitux - not sure how to get the _expected_ behavior either. I've tried on TT5 with `trunk` and I'm also unable to see the "check" icon being discussed.

Are we referring to the green "success" banner above the single product? Perhaps a quick video recording will help clarify.

**Comment by mattialerda** [comment]
Hi, we noticed that on sites where we updated to the latest version of WooCommerce the buttons styling has stopped working. On the “My Account” page we use the shortcode `[woocommerce_my_account]` and as you can see from the screenshot the styles are missing. By adding the `.woocommerce` class to the body the CSS selector works again. Is this the same issue as the ticket?

![Image](https://github.com/user-attachments/assets/a47abad7-f6fe-4022-a0e5-2872ff9e21b6)

**Comment by mikejolley** [comment]
@danielwrobert @gigitux Has something changed recently with body classes? I can no longer get the check mark to appear on block themes. I did previously I'm sure of it.

Related to this I guess:

- if I use a block theme, Twenty Twenty Five, I get the spinner on the shop page, but there is no spinner on the homepage buttons.
- if I used a non-block theme, say Twenty Twelve, I do get the check, but the button styling is now completely missing from the homepage. 

![Image](https://github.com/user-attachments/assets/6bbf2197-e18d-46b8-88ac-a35c05ba102e)

Its possible there is an expected difference between the block + shortcode experiences, but its hard to understand what powers what so to a merchant its unexpected/broken behaviour.

**Comment by imanish003** [comment]
I was able to reproduce the issue on Twenty Twelve theme. Here is a video showing the issue:

https://github.com/user-attachments/assets/3410d69f-6499-4dd0-9c9d-2653782c8e3d


I will continue looking into this and will update here with my findings.

**Comment by imanish003** [comment]
After testing on various WC versions, I can confirm that the issue wasn't present in and before WC 9.3.4. It was introduced in WC 9.4. I will continue looking into what change caused this.

**Comment by imanish003** [comment]
After investigation, I found out that [this PR](https://github.com/woocommerce/woocommerce/pull/51153) introduced the issue. 

**Before**
```css
.woocommerce:where(body:not(.woocommerce-block-theme-has-button-styles)),
:where(body:not(.woocommerce-block-theme-has-button-styles)) .woocommerce {
```



**After**
```css
.woocommerce:where(body:not(.woocommerce-block-theme-has-button-styles)),
:where(body:not(.woocommerce-block-theme-has-button-styles)):where(:not(.edit-post-visual-editor *)) .woocommerce
```

**Fix**

Just removing star `*` from `.edit-post-visual-editor` selector in the `:where()` function fix the issue. Is there any reason for having `*` there? @octaedro Can you please share reason for having `*` there? Is it okay if I can create a fix PR and remove `*`? 

---

Sidenote: I feel like these styles are a bit complicated and hard to maintain. For example:
- The first selector assumes that the `.woocommerce` class is present on the body tag. `.woocommerce` class added to the body tag [here](https://github.com/woocommerce/woocommerce/blob/2e0d04e80a30ecdf42b357a085061dc1f5a3f9a2/plugins/woocommerce/includes/wc-template-functions.php#L314-L319). This class is added only if `is_woocommerce` filter returns true i.e. on shop, product taxonomy or product pages. This explain the reason why styles were working on WooCommerce pages.
- OR as per 2nd Selector, Button should be nested in a element with a `.woocommerce` class i.e. `.woocommerce a.button`
	- Because of changes in [this PR](https://github.com/woocommerce/woocommerce/pull/51153), this stopped working.

It took me a good amount of time to even understand why the styles stopped working. This could be maintenance nightmare. Also, I agree with @mikejolley that it doesn't make sense to have style for class like ".woocommerce-block-theme-has-button-styles" inside `woocommerce.scss`. It should be moved to somewhere in WooCommerce blocks. OR near to button block itself. 

CC: @sunyatasattva What do you think about moving `.woocommerce-block-theme-has-button-styles` styles to WooCommerce blocks?
CC: @mdperez86 JFYI

**Comment by valentin-anamorphik** [comment]
Hi @imanish003 , 

Do you have more answers?
Is this a Woocommerce bug?

**Comment by imanish003** [comment]
@valentin-anamorphik Yes, based on my investigation this appears to be a WooCommerce bug. The CSS selector change in WC 9.4 is preventing button styles from being applied correctly on non-WooCommerce pages. While themes that override button styles with custom CSS may not be affected, this is still a regression in core WooCommerce functionality that should be fixed. I'll create a PR with a proposed fix to address this issue.

**Comment by imanish003** [comment]
Update: I have merged PR #53482, which contains the fix for this issue and will be available in WooCommerce 9.5.
